File: cpp/interest-is-interesting/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    
    const double required_for_rate_2 = 0;
    const double required_for_rate_3 = 1000;
    const double required_for_rate_4 = 5000;

    const double rate_1 = 3.213;
    const double rate_2 = 0.5;
    const double rate_3 = 1.621;
    const double rate_4 = 2.475;

    if(balance < required_for_rate_2)
        return rate_1;

    if(balance < required_for_rate_3)
        return rate_2;

    if(balance < required_for_rate_4)
        return rate_3;

    return rate_4;

}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {

    const double percent = 0.01;

    return balance * interest_rate( balance ) * percent;

}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {

    return balance + yearly_interest(balance);

}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {

    int needed_years = 0;
    double current_balance = balance;

    while( current_balance < target_balance ) {
        current_balance = annual_balance_update(current_balance);
        needed_years++;
    }

    return needed_years;
    
}

File: go/interest-is-interesting/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	switch {
	case balance < 0:
		return 3.213
	case balance < 1000:
		return 0.5
	case balance < 5000:
		return 1.621
	default:
		return 2.475
	}
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	return balance * float64(InterestRate(balance)) / 100
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
	var doYearsCalculation func(float64, float64, int) int
	doYearsCalculation = func(balance, targetBalance float64, years int) int {
		if balance >= targetBalance {
			return years
		} else {
			var nextBalance float64 = AnnualBalanceUpdate(balance)
			return doYearsCalculation(nextBalance, targetBalance, years+1)
		}
	}
	return doYearsCalculation(balance, targetBalance, 0)
}
